### ISSUE TITLE
Windows service ServiceBase access for pause/continue support

### ISSUE THREAD

**Issue by artiomchi** [dotnet/runtime#50021]
### Is your feature request related to a problem? Please describe.
Currently there's a simple way to run a netcore3 app as a windows service using the `.UseWindowsService()` extension. The issue is that this extension uses it's own `ServiceBase` class, which is the only class that receives pause and continue events, shutdown events and custom service commands.

### Describe the solution you'd like
Add a second generic extension method that allows a custom class to be passed to `.UseWindowsService<>()`, which will be based on the existing `WindowsServiceLifetime`, retaining all functionality, but allowing extra customisation.

### Describe alternatives you've considered
I've thought of altering the `WindowsServiceLifetime` class to receive a `ICollection<ServiceBase>` from DI, and register them alongside itself in the `ServiceBase.Run()` call, but that will mean the `ApplicationLifetime.ApplicationStarted` will be called only after the `WindowsServiceLifetime` service has started, not when all services have.

Also, it seems to make sense to only have one service registered, and move "service" functionality to hosted service classes instead.

### Additional context
I've added an implementation to my fork, branched off `release/3.0`, since I wasn't sure which branch would be best to start the fix from: [feature/windows-service-custom-class](https://github.com/artiomchi/Extensions/tree/feature/windows-service-custom-class), with the fix being in this commit: [d9bafd6
](https://github.com/artiomchi/Extensions/commit/d9bafd63027ba352af6839b0f0003a8e504eb519)

If you let me know whether I should do it from `master`, or any other branch, I can make a PR for this.

**Comment by RachelXGanon** [comment]
@artiomchi - I have the same issue. I need the OnSessionChange() which not exists in worker service.
Did you find a solution?
I'm using .NET 5
Thank you

**Comment by eerhardt** [comment]
I don't think we should duplicate all the APIs of `ServiceBase` in Windows Service Hosting's layer. It is straightforward to implement this yourself as illustrated above. Doing it that way, you have the full power of `ServiceBase` without needing to add additional APIs.

See also the conversation here: https://github.com/dotnet/runtime/issues/75276#issuecomment-1258446954

@RachelXGanon - have you tried the approach outlined above? Derive a class from `WindowsServiceLifetime` and register it as the `IHostLifetime` service. That approach is also described at: https://github.com/dotnet/runtime/issues/75276#issuecomment-1242336776

**Comment by RachelXGanon** [comment]
@eerhardt  I appreciate your quick response.
Does it mean that in case that I need a service with the `ServiceBase` functionality, I can create custom service that derived from `ServiceBase`, in console app project?
And I don't need the worker service template? 
Thank you.

**Comment by jjxtra** [comment]
I did the custom sub class but my service still cannot show a messagebox. Regardless of whether this is best practice or not, this did used to work with the old ServiceBase class. What needs to be done in .NET 6 to interact with the desktop?

**Comment by A9G-Data-Droid** [comment]
On the topic of what this adds; I looked at the source for [WindowsServiceLifetime](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs) and it doesn't implement `OnPause` at all. The windows service control does have a pause button. What happens when an admin presses pause in the service control dialogue? Nothing?

**Comment by snowkoan** [comment]
Doesn't this approach end up creating two WindowsServiceLifetime objects?

```
UseWindowsService() {
..
                    services.AddSingleton<IHostLifetime, WindowsServiceLifetime>();
..
}

```
And then a second one is added:

`services.AddSingleton<IHostLifetime, CustomService>();
`
It does seem to cause some problems, like intermittently not being able to mark the service stoppable, errors on shutdown. I'm not a .NET expert, so this is correlation at this point. But curious what others think.

And if it is a real issue, I could work around it by not calling UseWindowsService. But... it seems a bit silly to copy/paste that code into my CustomService.

**Comment by snowkoan** [comment]
Thanks, I appreciate the answer. This .NET stuff is another world.
